src/libs/select.c

```diff
@@ -220,30 +226,141 @@ static int lua_register_selection(lua_State *L)
   dt_lib_module_t *self = lua_touserdata(L, lua_upvalueindex(1));
   dt_lua_module_entry_push(L,"lib",self->plugin_name);
   lua_getuservalue(L,-1);
-  const char* key = luaL_checkstring(L,1);
-  luaL_checktype(L,2,LUA_TFUNCTION);
+  const char* name;
+  const char* key;
+  char* tmp_name = NULL;
+  // location of function argument on the stack: 2 - old format, 3 - new format (w/name)
+  int func_pos = 2;
+
+  if (lua_type(L, 3) == LUA_TFUNCTION)
+  {
+    name = luaL_checkstring(L,1);
+    key = luaL_checkstring(L,2);
+    func_pos = 3;
+  }
+  else
+  {
+    key = luaL_checkstring(L,1);
+    tmp_name = malloc(sizeof(key));
+    // replace spaces in name with underscores
+    for(int i = 0; key[i] != '\0'; i++)
+      if(key[i] == ' ')
+        tmp_name[i] = '_';
+      else
+        tmp_name[i] = key[i];
+
+    name = tmp_name;
+  }
+
+  luaL_checktype(L,func_pos,LUA_TFUNCTION);
 
   lua_getfield(L,-1,"callbacks");
-  lua_pushstring(L,key);
-  lua_pushvalue(L,2);
+  lua_pushstring(L,name);
+  lua_pushvalue(L,func_pos);
   lua_settable(L,-3);
 
   GtkWidget* button = gtk_button_new_with_label(key);
   const char * tooltip = lua_tostring(L,3);
   if(tooltip)  {
     gtk_widget_set_tooltip_text(button, tooltip);
   }
+  gtk_widget_set_name(button, name);
   gtk_grid_attach_next_to(GTK_GRID(self->widget), button, NULL, GTK_POS_BOTTOM, 2, 1);
 
 
   lua_callback_data * data = malloc(sizeof(lua_callback_data));
-  data->key = strdup(key);
+  data->key = strdup(name);
   data->self = self;
-  g_signal_connect(G_OBJECT(button), "clicked", G_CALLBACK(lua_button_clicked), data);
+  gulong s = g_signal_connect(G_OBJECT(button), "clicked", G_CALLBACK(lua_button_clicked), data);
+
+  dt_lua_module_entry_push(L,"lib",self->plugin_name);
+  lua_getuservalue(L,-1);
+  lua_getfield(L,-1,"signal_handlers");
+  lua_pushstring(L,name);
+  lua_pushinteger(L,s);
+  lua_settable(L,-3);
+
   gtk_widget_show_all(self->widget);
+
+  if(tmp_name){
```

Comment:
{ should be on next line. But in this case just:

`if(tmp_name) free(tmp_name);`

Reply:
removed